### PR TITLE
A-case: TransactionStepView

### PR DESCRIPTION
**Pull request by robinsalehjan**
# Why?

- Need a vertical step indicator along with N steps in order to facilitate for the journey of a car transaction.

# What?

- This is still a work in progress 
- Would like to have it in master and iterate over it since some components are missing and needs dark mode tweaks.

# Show me

| iPhone | iPad |
| --- | --- |
| Simulator Screen Shot - iPhone 11 Pro Max - 2020-02-25 at 10 31 23 | Simulator Screen Shot - iPad Pro (12 9-inch) (3rd generation) - 2020-02-25 at 10 32 26 |

### CODE REVIEW

File: Sources/Fullscreen/TransactionView/TransactionStepDot.swift [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+public class TransactionStepDot: UIView {
+    enum State {
+        case inProgress
+        case completed
+    }
+
+    // MARK: - Public properties
+
+    public static let activeColor = UIColor(r: 82, g: 188, b: 245)
```
Comment (ag12) [inline Sources/Fullscreen/TransactionView/TransactionStepDot.swift]: Are these Dark mode compatible?

File: Sources/Fullscreen/TransactionView/TransactionStepDot.swift [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        let innerSize = CABasicAnimation(keyPath: "transform.scale")
+        innerSize.isRemovedOnCompletion = false
+        innerSize.fillMode = .forwards
+
+        switch state {
+        case .inProgress:
+            outerFill.toValue = TransactionStepDot.activeColor.cgColor
+            innerSize.toValue = [ 1.0, 1.0 ]
+        case .completed:
+            outerFill.toValue = TransactionStepDot.activeColor.cgColor
```
Comment (ag12) [inline Sources/Fullscreen/TransactionView/TransactionStepDot.swift]: Do you want to use `activeColor.cgColor` for both inProgress and completed?

File: Sources/Fullscreen/TransactionView/TransactionStepView/TransactionStepView.swift [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+    private lazy var titleView: UITextView = {
+        let view = UITextView(withAutoLayout: true)
+        view.font = style.titleFont
+        view.textColor = style.titleTextColor
+        view.backgroundColor = style.backgroundColor
+        view.isScrollEnabled = false
+        view.isEditable = false
+        view.contentInset = .init(top: -.mediumSpacing, leading: 0, bottom: 0, trailing: 0)
+        view.translatesAutoresizingMaskIntoConstraints = false
```
Comment (ag12) [inline Sources/Fullscreen/TransactionView/TransactionStepView/TransactionStepView.swift]: is `view.translatesAutoresizingMaskIntoConstraints = false` needed if you are using `UITextView(withAutoLayout: true)`

File: Sources/Fullscreen/TransactionView/TransactionStepView/TransactionStepView.swift [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+    private lazy var bodyView: UITextView = {
+        let view = UITextView(withAutoLayout: true)
+        view.font = style.bodyFont
+        view.textColor = style.bodyTextColor
+        view.backgroundColor = style.backgroundColor
+        view.isScrollEnabled = false
+        view.isEditable = false
+        view.contentInset = .init(top: -.mediumSpacing, leading: 0, bottom: 0, trailing: 0)
+        view.translatesAutoresizingMaskIntoConstraints = false
```
Comment (ag12) [inline Sources/Fullscreen/TransactionView/TransactionStepView/TransactionStepView.swift]: +1

File: Sources/Fullscreen/TransactionView/TransactionStepView/TransactionStepView.swift [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+    private lazy var detailView: UITextView = {
+        let view = UITextView(withAutoLayout: true)
+        view.font = style.detailFont
+        view.textColor = style.detailTextColor
+        view.backgroundColor = style.backgroundColor
+        view.isScrollEnabled = false
+        view.isEditable = false
+        view.contentInset = .leadingInset(0)
+        view.translatesAutoresizingMaskIntoConstraints = false
```
Comment (ag12) [inline Sources/Fullscreen/TransactionView/TransactionStepView/TransactionStepView.swift]: 😅

File: Sources/Fullscreen/TransactionView/TransactionView.swift [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+            scrollableContentView.heightAnchor.constraint(greaterThanOrEqualTo: scrollView.heightAnchor, constant: .mediumSpacing),
+            scrollableContentView.widthAnchor.constraint(equalTo: scrollView.widthAnchor),
+
+            verticalStackView.trailingAnchor.constraint(equalTo: scrollableContentView.trailingAnchor, constant: -.mediumLargeSpacing),
+            verticalStackView.topAnchor.constraint(equalTo: scrollableContentView.topAnchor, constant: .largeSpacing),
+            verticalStackView.bottomAnchor.constraint(equalTo: scrollableContentView.bottomAnchor),
+        ])
+
+        for index in 0..<numberOfSteps {
```
Comment (ag12) [inline Sources/Fullscreen/TransactionView/TransactionView.swift]: would be nice with space after <?

File: Sources/Fullscreen/TransactionView/TransactionView.swift [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+            guard let model = dataSource?.transactionViewModelForIndex(self, forStep: index) else { return }
+
+            addTransactionStepView(index, model)
+            addTransactionStepDots(index)
+        }
+
+        guard let stepDot = stepDots.first else { return }
+        verticalStackView.leadingAnchor.constraint(equalTo: stepDot.trailingAnchor, constant: .mediumSpacing).isActive = true
```
Comment (ag12) [inline Sources/Fullscreen/TransactionView/TransactionView.swift]: why not just use `NSLayoutConstraint.activate([])`

File: Sources/Fullscreen/TransactionView/TransactionStepDot.swift (hunk 1 above)
Comment (robinsalehjan) [reply Sources/Fullscreen/TransactionView/TransactionStepDot.swift]: Nope, will be changed in next iteration

File: Sources/Fullscreen/TransactionView/TransactionStepDot.swift (hunk 2 above)
Comment (robinsalehjan) [reply Sources/Fullscreen/TransactionView/TransactionStepDot.swift]: For now yes, but may change.

File: Sources/Fullscreen/TransactionView/TransactionStepView/TransactionStepView.swift (hunk 3 above)
Comment (robinsalehjan) [reply Sources/Fullscreen/TransactionView/TransactionStepView/TransactionStepView.swift]: Will fix 👍

File: Sources/Fullscreen/TransactionView/TransactionView.swift (hunk 7 above)
Comment (robinsalehjan) [reply Sources/Fullscreen/TransactionView/TransactionView.swift]: It's just one anchor, but could use it I guess.

Comment (ag12) [reply Sources/Fullscreen/TransactionView/TransactionView.swift]: most of my comments are minor/ trivial :D
most == all 💯

**Pull request merged**